Source/Particles/Deposition/CurrentDeposition.H

```diff
@@ -699,181 +699,204 @@ void doVayDepositionShapeN (const GetParticlePosition& GetPosition,
+        // Keep these double to avoid bug in single precision.
+        double sy_new[depos_order+1] = {0.};
+        double sy_old[depos_order+1] = {0.};
 #endif
-        amrex::Real sz_new[depos_order+1] = {0._rt};
-        amrex::Real sz_old[depos_order+1] = {0._rt};
+        // Keep these double to avoid bug in single precision.
+        double sz_new[depos_order+1] = {0.};
+        double sz_old[depos_order+1] = {0.};
 
         // Compute shape factors for current positions
 
         // i_new leftmost grid point in x that the particle touches
         // sx_new shape factor along x for the centering of each current
-        const int i_new = compute_shape_factor<depos_order>(sx_new, x_new);
+        Compute_shape_factor< depos_order > const compute_shape_factor;
+        const int i_new = compute_shape_factor(sx_new, x_new);
 #if (defined WARPX_DIM_3D)
         // j_new leftmost grid point in y that the particle touches
         // sy_new shape factor along y for the centering of each current
-        const int j_new = compute_shape_factor<depos_order>(sy_new, y_new);
+        const int j_new = compute_shape_factor(sy_new, y_new);
 #endif
         // k_new leftmost grid point in z that the particle touches
         // sz_new shape factor along z for the centering of each current
-        const int k_new = compute_shape_factor<depos_order>(sz_new, z_new);
+        const int k_new = compute_shape_factor(sz_new, z_new);
 
         // Compute shape factors for old positions
 
         // i_old leftmost grid point in x that the particle touches
         // sx_old shape factor along x for the centering of each current
-        const int i_old = compute_shape_factor<depos_order>(sx_old, x_old);
+        const int i_old = compute_shape_factor(sx_old, x_old);
 #if (defined WARPX_DIM_3D)
         // j_old leftmost grid point in y that the particle touches
         // sy_old shape factor along y for the centering of each current
-        const int j_old = compute_shape_factor<depos_order>(sy_old, y_old);
+        const int j_old = compute_shape_factor(sy_old, y_old);
 #endif
         // k_old leftmost grid point in z that the particle touches
         // sz_old shape factor along z for the centering of each current
-        const int k_old = compute_shape_factor<depos_order>(sz_old, z_old);
+        const int k_old = compute_shape_factor(sz_old, z_old);
 
         // Deposit current into jx_arr, jy_arr and jz_arr
-
 #if (defined WARPX_DIM_XZ)
 
         for (int k=0; k<=depos_order; k++) {
             for (int i=0; i<=depos_order; i++) {
 
+                // Re-casting sx_new and sz_new from double to amrex::Real so that
+                // Atomic::Add has consistent types in its argument
+                amrex::Real xnzn = sx_new[i] * sz_new[k];
+                amrex::Real xozn = sx_old[i] * sz_new[k];
+                amrex::Real xnzo = sx_new[i] * sz_old[k];
+                amrex::Real xozo = sx_old[i] * sz_old[k];
```

Comment:
Here and in the following, I would personally prefer to keep the letter "s" in the name of these variables, in order to emphasize the fact that they represent splines (or shape factors) along the different directions. I also tried to separate logically distinct code blocks with blank lines when I wrote this part of the code, as the implementation is not that easy to follow in the first place, so I added a blank line as well after the code blocks where the new variables are defined, in order to separate them from the `for` loops and the assignments of the various components of `J`.

```suggestion
                amrex::Real const sxn_szn = sx_new[i] * sz_new[k];
                amrex::Real const sxo_szn = sx_old[i] * sz_new[k];
                amrex::Real const sxn_szo = sx_new[i] * sz_old[k];
                amrex::Real const sxo_szo = sx_old[i] * sz_old[k];
                
```

Reply:
Sounds good, and `const`-ified (edited).